### PR TITLE
adminHome.js does not URL encode form data before passing it on as GET params #6397

### LINKED ISSUE #6397: adminHome.js does not URL encode form data before passing it on as GET params
Form data is usually encoded by the browser automatically upon submit for HTML `form` elements. However, as AJAX is used to add instructors, custom methods `addInstructorFromFirstFormByAjax()` and `addInstructorFromSecondFormByAjax()` are used to read raw values from the form and construct the parameter query string used in the GET request to `/admin/adminInstructorAccountAdd`. Currently, the raw values are used directly ([here](https://github.com/TEAMMATES/teammates/blob/7a2eb84eda7105548d7d13cdaf5065778f7468f0/src/main/webapp/js/adminHome.js#L127) and [here](https://github.com/TEAMMATES/teammates/blob/7a2eb84eda7105548d7d13cdaf5065778f7468f0/src/main/webapp/js/adminHome.js#L149)) to form the query string, without going through URL encoding. If the data contains special characters, these will be misinterpreted.

Example of an undesirable effect: unable to add instructor with email address containing `+`, as `+` is interpreted as a space in the email address:
screen shot 2017-01-20 at 2 07 49 pm

### PR DESCRIPTION
**Pull request by thenaesh**
Fixes #6397

### CODE REVIEW

**Comment by thenaesh** [comment]
@whipermr5 Unfortunately, encoding the whole email address causes the email address to fail the email address validation (mainly because `@` gets encoded). I'm looking into fixing the validation code so that this becomes a non-issue.

**Comment by whipermr5** [comment]
It's not the email address validation that's failing, but the course ID validation. Upon adding of a new instructor, a demo course is generated based on the instructor email, but the generated course ID imports the `+` from the email address (see #6398). I've fixed #6398 locally and indeed, encoding the entire email address works perfectly.

**Comment by thenaesh** [comment]
Oh ok, that makes sense now. Thanks!

**Comment by thenaesh** [comment]
Ready to review.

File: src/main/webapp/js/adminHome.js [hunk 1]
```diff
@@ -30,6 +30,41 @@ function createRowForResultTable(shortName, name, email, institution, isSuccess,
...
+ * e.g.
+ * 'instructorname=Test Instructor&instructoremail=test.instructor@gmail.com'
+ * becomes
+ * 'instructorname=Test%20Instructor&instructoremail=test.instructor%40gmail.com'
+ *
+ * @param queryString
+ * @returns the encoded query string
+ */
+function encodeQueryString(queryString) {
+    return queryString.split('&')
```
Comment (whipermr5) [inline src/main/webapp/js/adminHome.js]: What if the user input contains `&`? I feel the encoding should happen before the querystring is constructed

File: src/main/webapp/js/adminHome.js [hunk 2]
```diff
@@ -145,12 +180,12 @@ function addInstructorFromSecondFormByAjax() {
     
     var instructorDetails = $('#instructorName').val() + '|' + $('#instructorEmail').val()
                             + '|' + $('#instructorInstitution').val();
-    instructorDetailsList = [instructorDetails];
+    instructorDetailsList = [encodeInstructorDetails(instructorDetails)];
     var params = 'instructorshortname=' + $('#instructorShortName').val()
```
Comment (whipermr5) [inline src/main/webapp/js/adminHome.js]: I.e. `var params = 'instructorshortname=' + encodeURIComponent($('#instructorShortName').val())` etc

File: src/main/webapp/js/adminHome.js (hunk 1 above)
Comment (whipermr5) [inline src/main/webapp/js/adminHome.js]: What if the user input contains `&`? The data should be encoded before constructing the query string. Or you could use [`jQuery.param`](http://api.jquery.com/jQuery.param/) to do both together

Comment (thenaesh) [inline src/main/webapp/js/adminHome.js]: Oh yea, good point.

File: src/main/webapp/js/adminHome.js (hunk 2 above)
Comment (thenaesh) [inline src/main/webapp/js/adminHome.js]: Just had a look, `$.param` looks like it actually encodes the string properly, so all the separate encoding functions are unnecessary. I'll modify that part to use a query object then.

Comment (whipermr5) [inline src/main/webapp/js/adminHome.js]: Yup 👍👍

**Comment by thenaesh** [comment]
@whipermr5 I've made the changes you suggested. I removed the extra functions at the top and just used `encodeURIComponent` directly. Also, `$.param` makes the code look nicer I think, no more + signs all over the place. xD

If you have any other suggestions, do let me know.

**Comment by whipermr5** [comment]
Nice! Looks good to me :)

**Comment by thenaesh** [comment]
Two issues:
1) I think it is better to wait for #6408 to be merged, since the error messages I use in the test will change.
2) I just discovered that this PR will open up a latent vulnerability, which I'm trying to fix first.

**Comment by thenaesh** [comment]
@wkurniawan07 Done! Ready for review.

**Comment by whipermr5** [comment]
Looks good! Without your changes, the updated test case would fail as `+` would be interpreted as a space.

File: src/test/java/teammates/test/cases/browsertests/AdminHomePageUiTest.java [hunk 3]
```diff
@@ -74,14 +74,24 @@ private void testCreateInstructorAction() throws Exception {
...
+        String demoCourseId = "AHPUiT____.instr1_.gma-demo";
         
         String instructorDetails = instructor.name + " | " + instructor.email + "\n"
                                  + instructor.name + " | " + instructor.email + " | " + institute;
         
+        ______TS("action success: instructor details are sent across to the backend properly");
+        BackDoor.deleteAccount(TestProperties.TEST_INSTRUCTOR_ACCOUNT);
+        BackDoor.deleteCourse(demoCourseId);
+        BackDoor.deleteInstructor(demoCourseId, instructor.email);
+        homePage.createInstructorByInstructorDetailsSingleLineForm(instructorDetails);
```
Comment (wkurniawan07) [inline src/test/java/teammates/test/cases/browsertests/AdminHomePageUiTest.java]: What's the reason for this new sequence of action? From what I understand, the two `assertEquals` below can be moved to the pre-existing test sequence immediately after this one.

Comment (thenaesh) [inline src/test/java/teammates/test/cases/browsertests/AdminHomePageUiTest.java]: Oh yea you're right. I initially used a different set of instructor details for this test alone, so I had to do this. Then I realised that it wasn't necessary to use a different set of details and removed those, but forgot to remove this. I'll fix it. Thanks for pointing it out.

**Comment by thenaesh** [comment]
@wkurniawan07 I've removed that duplicate code.

File: src/test/java/teammates/test/cases/browsertests/AdminHomePageUiTest.java [hunk 4]
```diff
@@ -74,19 +74,25 @@ private void testCreateInstructorAction() throws Exception {
...
+        instructor.email = "AHPUiT+++_.instr1!@gmail.tmt";
         String institute = "TEAMMATES Test Institute 1";
-        String demoCourseId = "AHPUiT.instr1_.gma-demo";
+        String demoCourseId = "AHPUiT____.instr1_.gma-demo";
         
         String instructorDetails = instructor.name + " | " + instructor.email + "\n"
                                  + instructor.name + " | " + instructor.email + " | " + institute;
         
-        ______TS("action fail & success: add multiple instructors");
+        ______TS("action success: instructor details are sent across to the backend properly");
```
Comment (wkurniawan07) [inline src/test/java/teammates/test/cases/browsertests/AdminHomePageUiTest.java]: It seems like there are unnecessary changes now, like this.

File: src/test/java/teammates/test/cases/browsertests/AdminHomePageUiTest.java [hunk 5]
```diff
@@ -74,19 +74,25 @@ private void testCreateInstructorAction() throws Exception {
...
         BackDoor.deleteAccount(TestProperties.TEST_INSTRUCTOR_ACCOUNT);
         BackDoor.deleteCourse(demoCourseId);
         BackDoor.deleteInstructor(demoCourseId, instructor.email);
         homePage.createInstructorByInstructorDetailsSingleLineForm(instructorDetails);
+        InstructorAttributes instructorInBackend = BackDoor.getInstructorByEmail(instructor.email, demoCourseId);
+        
+        assertEquals(instructor.getName(), instructorInBackend.getName());
+        assertEquals(instructor.getEmail(), instructorInBackend.getEmail());
+        
+        ______TS("action fail & success: add multiple instructors");
```
Comment (wkurniawan07) [inline src/test/java/teammates/test/cases/browsertests/AdminHomePageUiTest.java]: And this as well. This is why my previous suggestion was to move the two `assertEquals`. To be more precise, to between the `assertEquals("Instructor AHPUiT Instructor+++...` and `homePage.clearInstructorDetailsSingleLineForm();`.

**Comment by thenaesh** [comment]
@wkurniawan07 Moved the `assert`s to between the existing test sequence as per your recommendations. I removed the test segment I added earlier though, as it looks weird in that location. xD

File: src/test/java/teammates/test/cases/browsertests/AdminHomePageUiTest.java [hunk 6]
```diff
@@ -74,10 +74,10 @@ private void testCreateInstructorAction() throws Exception {
         InstructorAttributes instructor = new InstructorAttributes();
         
         String shortName = "Instrúctör";
-        instructor.name = "AHPUiT Instrúctör";
-        instructor.email = "AHPUiT.instr1!@gmail.tmt";
+        instructor.name = "AHPUiT Instrúctör+++";
```
Comment (damithc) [inline src/test/java/teammates/test/cases/browsertests/AdminHomePageUiTest.java]: Can this made more realistic? e.g. `AHPUiT Instrúctör WithPlusInEmail` Names don't have `+` in them. Unrealistic data makes the test outputs look unrealistic, making it hard to detect rubbish output caused by bugs.

**Comment by thenaesh** [comment]
@damithc Changes made.

File: src/test/java/teammates/test/cases/browsertests/AdminHomePageUiTest.java [hunk 7]
```diff
@@ -74,10 +74,10 @@ private void testCreateInstructorAction() throws Exception {
         InstructorAttributes instructor = new InstructorAttributes();
         
         String shortName = "Instrúctör";
-        instructor.name = "AHPUiT Instrúctör";
-        instructor.email = "AHPUiT.instr1!@gmail.tmt";
+        instructor.name = "AHPUiT Instrúctör WithNameInEmail";
```
Comment (damithc) [inline src/test/java/teammates/test/cases/browsertests/AdminHomePageUiTest.java]: `WithNameInEmail` doesn't indicate that I should expect weird chars in email address of this guy. Furthermore, most test data follow the pattern that user name is partially/fully included in the email address, so there is no reason to mention it in just this test case. Either omit it or use a more meaningful 'name'?

Comment (thenaesh) [inline src/test/java/teammates/test/cases/browsertests/AdminHomePageUiTest.java]: Oops, typo.
